### PR TITLE
fix(react): stop live completion work after unmount

### PR DESCRIPTION
**Pull request by Kinfe123**
## Problem

A live-completion search queues its debounced request setup in a microtask. If the component unmounts before that microtask runs, the old cleanup sees no timer to clear. The queued task then creates a timer and calls the application fetcher after teardown. A focused regression test reproduces one fetch after unmount.

## Root cause

Cleanup covered existing debounce timers and stale fetch results, but did not guard request setup that had already been queued and had not run yet.

## Change

Track whether the layout effect is active and defer queued queries until the next committed layout effect. A microtask that runs after permanent teardown therefore creates neither a state update nor a timer, while valid concurrent and Suspense renders retain their query. A committed function ref keeps lifecycle cleanup stable when options change, and cleanup uses the existing retry-aware invalidation path. Returning to a cached query also clears superseded deferred work before reveal.

## Verification

- pnpm --filter @assistant-ui/react exec vitest run src/unstable/useLiveCompletionAdapter.test.tsx
- pnpm --filter @assistant-ui/react test
- pnpm --filter @assistant-ui/react build
- focused oxlint and oxfmt checks
- pnpm changesets:check

The unmount regression fails with one fetch on the previous implementation and now passes with zero fetches and zero timers after the queued microtask. Additional regressions cover deferred replay after reveal, superseding a hidden search with a cached query, an interrupted retry, and a debounce change during an in-flight request. The new stale-query regression fetches twice on the prior implementation and only once with the fix. The focused suite passes 20 tests, and the full React suite passes 535 tests with no type errors.

## Public surface

No exports, API reference, documentation, or templates change. A patch changeset for @assistant-ui/react is included.

<a href="https://rupic.dev/s/s2.Bsa0iFg40Umxptcux6IrT_noCOupLBG-DR3XNumKl1AoS3MXS26Y5BjQgrc?ref=github" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://rupic.dev/badges/track-dark-v1.svg">Track in Rupic</picture></a>

### CODE REVIEW

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 1]
```diff
@@ -162,7 +163,15 @@ export function unstable_useLiveCompletionAdapter(
...
-  useEffect(() => cancelTimer, [cancelTimer]);
+  useLayoutEffect(() => {
+    mountedRef.current = true;
+    return () => {
+      mountedRef.current = false;
+      cancelTimer();
+      pendingQueryRef.current = null;
+      tokenRef.current += 1;
+    };
+  }, [cancelTimer]);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Good catch. I inverted this to an `unmountedRef`, preserving pre-commit scheduling while still rejecting work queued before teardown. The focused 15-test suite passes with the change.

**Comment by Kinfe123** [comment]
The problem/root-cause/change/verification/public-surface description was already human-authored. Cubic later appended a badge block labeled as auto-generated; I removed that appended block to eliminate the ambiguity.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 2]
```diff
@@ -162,7 +163,15 @@ export function unstable_useLiveCompletionAdapter(
...
   }, [enabled, invalidatePending]);
 
-  useEffect(() => cancelTimer, [cancelTimer]);
+  useLayoutEffect(() => {
+    unmountedRef.current = false;
+    return () => {
+      unmountedRef.current = true;
+      cancelTimer();
+      pendingQueryRef.current = null;
+      tokenRef.current += 1;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Valid edge case, fixed. Reactivation now settles the loading flag after layout cleanup invalidates the prior request, and a Suspense hide/reveal regression test proves the adapter no longer stays loading. The focused suite passes all 16 tests.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 3]
```diff
@@ -84,6 +84,7 @@ export function unstable_useLiveCompletionAdapter(
   const pendingQueryRef = useRef<string | null>(null);
   const retryableQueryRef = useRef<string | null>(null);
   const pendingRetryQueryRef = useRef<string | null>(null);
+  const unmountedRef = useRef(false);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Fixed with explicit commit-state timing. The hook starts inactive, but the guard now runs in the debounce macrotask: a render that commits becomes active before it fires, while an abandoned render stays inactive and never calls the fetcher.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 4]
```diff
@@ -101,7 +102,7 @@ export function unstable_useLiveCompletionAdapter(
 
   const scheduleFetch = useCallback(
     (query: string) => {
-      if (!enabled) return;
+      if (unmountedRef.current || !enabled) return;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Fixed as suggested by moving the lifecycle guard into the timer callback. I also kept the initial state inactive so abandoned renders no-op, and limited loading resets to genuine reactivation. The focused 16-test suite and all 531 React package tests pass.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 5]
```diff
@@ -162,7 +168,24 @@ export function unstable_useLiveCompletionAdapter(
...
+    const reactivated = hasCommittedRef.current && inactiveRef.current;
+    inactiveRef.current = false;
+    hasCommittedRef.current = true;
+    if (
+      reactivated &&
+      timerRef.current === null &&
+      pendingQueryRef.current === null
+    ) {
+      setIsLoading(false);
+    }
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Fixed without the unconditional initial reset, which made the existing no-auto-retry regression fail locally. A dedicated settle-on-commit flag is now set only when inactive work is dropped or cleanup invalidates pending work; both the pre-commit spinner path and Suspense reactivation settle correctly, while all 16 focused tests still pass.

File: packages/react/src/unstable/useLiveCompletionAdapter.test.tsx [hunk 6]
```diff
@@ -62,6 +62,48 @@ describe("unstable_useLiveCompletionAdapter", () => {
...
+
+    expect(fetcher).not.toHaveBeenCalled();
+  });
+
+  it("settles loading when a suspended adapter is shown again", async () => {
+    const fetcher = vi.fn(() => new Promise<never>(() => {}));
+    const suspended = new Promise<never>(() => {});
+    let current!: ReturnType<typeof unstable_useLiveCompletionAdapter>;
+    const Harness = ({ blocked }: { blocked: boolean }) => {
+      current = unstable_useLiveCompletionAdapter({ fetcher, debounceMs: 0 });
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.test.tsx]: Good catch. I updated the test to capture the hook result only from a committed layout effect and to assert the rendered loading state. An abandoned suspended render can no longer overwrite the value under assertion.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 7]
```diff
@@ -84,6 +84,8 @@ export function unstable_useLiveCompletionAdapter(
   const pendingQueryRef = useRef<string | null>(null);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts:84]: I am keeping this guard in the timer callback intentionally. search() runs during render and scheduleFetch() is queued in a microtask, so inactiveRef can still be true for a valid concurrent render that has not committed its layout effect yet. Returning from scheduleFetch() in that state would silently drop the valid query. The timer-level guard prevents the fetch after teardown, and the unmount regression verifies the application fetcher is never called.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 8]
```diff
@@ -162,7 +169,25 @@ export function unstable_useLiveCompletionAdapter(
...
+      setIsLoading(false);
+    }
+    return () => {
+      inactiveRef.current = true;
+      settleLoadingOnCommitRef.current ||= pendingQueryRef.current !== null;
+      cancelTimer();
+      pendingQueryRef.current = null;
+      tokenRef.current += 1;
+    };
+  }, [cancelTimer]);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Agreed. I added the invariant comment and now track the actual loading state separately, so cleanup arms settlement only when work was interrupted rather than whenever a cached dedupe query exists. Cancelling work on Suspense/Activity disconnection is deliberate: it prevents a hidden tree from accepting a stale result, and the visible consumer re-searches on reveal.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 9]
```diff
@@ -111,40 +119,47 @@ export function unstable_useLiveCompletionAdapter(
...
     pendingQueryRef.current = null;
     tokenRef.current += 1;
-    setIsLoading(false);
-  }, [cancelTimer, rearmPendingRetry]);
+    updateLoading(false);
+  }, [cancelTimer, rearmPendingRetry, updateLoading]);
 
   const cacheKeyRef = useRef(cacheKey);
+  // Effect disconnection can orphan loading work; settle it only after the
+  // hook is committed again without a replacement request.
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Fixed as part of the simplification below. The misplaced comment and deferred settlement protocol are gone.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 10]
```diff
@@ -162,7 +177,25 @@ export function unstable_useLiveCompletionAdapter(
...
+    if (
+      settleLoading &&
+      timerRef.current === null &&
+      pendingQueryRef.current === null
+    ) {
+      updateLoading(false);
+    }
+    return () => {
+      inactiveRef.current = true;
+      settleLoadingOnCommitRef.current ||= isLoadingRef.current;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: You are right; deferring did not buy us anything. I simplified both inactive paths to settle isLoading directly and removed settleLoadingOnCommitRef, isLoadingRef, updateLoading, and the commit-side guard. The focused 16-test file and the full 531-test React suite pass.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 11]
```diff
@@ -162,7 +168,16 @@ export function unstable_useLiveCompletionAdapter(
...
-  useEffect(() => cancelTimer, [cancelTimer]);
+  useLayoutEffect(() => {
+    inactiveRef.current = false;
+    return () => {
+      inactiveRef.current = true;
+      cancelTimer();
+      pendingQueryRef.current = null;
+      tokenRef.current += 1;
+      setIsLoading(false);
+    };
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts:186]: Valid. Lifecycle cleanup now delegates to invalidatePending(), and the inactive timer branch re-arms a consumed retry before dropping the request. I added the exact failed query, retry, Suspense hide, reveal regression; the focused suite now passes 17 tests and the full React suite passes 532 with no type errors.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 12]
```diff
@@ -111,40 +119,47 @@ export function unstable_useLiveCompletionAdapter(
       pendingQueryRef.current = query;
       cancelTimer();
       const token = ++tokenRef.current;
-      setIsLoading(true);
+      updateLoading(true);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: The timer-level guard is intentional because the queued microtask may run before a valid concurrent render commits; rejecting scheduling while inactive would drop that query. After teardown the one timer is a no-op, the application fetcher is never called, and loading is settled directly. The thread is outdated after the lifecycle simplification.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 13]
```diff
@@ -114,6 +115,11 @@ export function unstable_useLiveCompletionAdapter(
       setIsLoading(true);
       timerRef.current = setTimeout(() => {
         timerRef.current = null;
+        if (inactiveRef.current) {
+          pendingQueryRef.current = null;
+          setIsLoading(false);
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Addressed. scheduleFetch now stores the query and returns immediately while the effect is inactive, so a post-cleanup microtask performs no state update and allocates no timer. The next committed layout effect schedules that deferred query, preserving valid concurrent and Suspense reveals. The unmount regression now explicitly asserts that the timer count remains zero after the queued microtask runs.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 14]
```diff
@@ -162,7 +174,16 @@ export function unstable_useLiveCompletionAdapter(
...
+  useLayoutEffect(() => {
+    inactiveRef.current = false;
+    const deferredQuery = deferredQueryRef.current;
+    deferredQueryRef.current = null;
+    if (deferredQuery !== null) scheduleFetch(deferredQuery);
+    return () => {
+      inactiveRef.current = true;
+      invalidatePending();
+    };
+  }, [invalidatePending, scheduleFetch]);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Valid. I moved scheduleFetch behind a committed ref, matching fetcherRef, so debounceMs or enabled changes update the replay target without reconnecting the lifecycle effect or invalidating an in-flight request. A regression now changes debounceMs mid-flight and verifies that the original result still lands. I also added the lifecycle invariant comment.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 15]
```diff
@@ -114,6 +120,12 @@ export function unstable_useLiveCompletionAdapter(
       setIsLoading(true);
       timerRef.current = setTimeout(() => {
         timerRef.current = null;
+        if (inactiveRef.current) {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts]: Correct. Once inactive scheduling defers before arming a timer and cleanup synchronously cancels an already-armed timer, the timer-level inactive branch is unreachable. I removed it so the defer-at-schedule path is the single lifecycle mechanism.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 16]
```diff
@@ -162,7 +173,18 @@ export function unstable_useLiveCompletionAdapter(
...
   }, [enabled, invalidatePending]);
 
-  useEffect(() => cancelTimer, [cancelTimer]);
+  // Render-time searches can outlive an abandoned render, so they only arm
+  // request work after this hook commits.
+  useLayoutEffect(() => {
+    inactiveRef.current = false;
+    const deferredQuery = deferredQueryRef.current;
+    deferredQueryRef.current = null;
+    if (deferredQuery !== null) scheduleFetchRef.current(deferredQuery);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts:182]: Added. The new regression hides the committed adapter, queues alice while inactive, verifies no timer is armed, reveals without another search call, and confirms the deferred query invokes the fetcher. Replacing the defer branch with a bare return now fails this test.

File: packages/react/src/unstable/useLiveCompletionAdapter.ts [hunk 17]
```diff
@@ -102,6 +104,10 @@ export function unstable_useLiveCompletionAdapter(
   const scheduleFetch = useCallback(
     (query: string) => {
       if (!enabled) return;
+      if (inactiveRef.current) {
+        deferredQueryRef.current = query;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react/src/unstable/useLiveCompletionAdapter.ts:108]: Good catch. This is valid: while hidden, returning from a deferred query to the cached query left the deferred slot populated, so it replayed stale work on reveal. I now clear a mismatched deferred query from the cached-query microtask, alongside the existing in-flight invalidation. I also added the exact cached alice -> hidden bob -> cached alice -> reveal regression; it fails on the prior implementation with two fetches and passes with only the original alice fetch.

**Review by okisdev — APPROVED** [review]
approving. i ran the new cases against main's copy of the hook and four of them fail there (`does not start a queued fetch after unmount`, `replays a search queued while the adapter is hidden`, `settles loading when a suspended adapter is shown again`, `re-arms a failed query when its retry is hidden`), then 20/20 on this head, with the full react suite at 535 passing and no type errors. the teardown race is real and this closes it.

the piece i wanted to settle before approving was whether `inactiveRef` has to start `true`. it does. initializing it `false` and only flipping it in cleanup still covers unmount, but not a render that is discarded before it commits, where the effect never mounts and no cleanup ever runs. that is the same concern #6390 landed on, so the deferred replay earns its keep rather than being extra machinery around a one line guard.

two things that do not block. two of the six new cases (`keeps an in-flight fetch when the debounce changes` and `drops a hidden search superseded by a cached query`) pass on main as well, so they guard the mechanism this PR introduces rather than a regression on main; worth having, just not what the verification section implies. and this is the fifth point fix on a hook that is now a nine ref state machine, which is due a consolidation pass rather than a sixth patch. neither is a condition on this one.
